impy/models/sophia.py

```diff
@@ -5,158 +5,191 @@
+            info(1, 'Ignores K0. Use K0L/S 130/310 in final state definition.')
+            return
+        idb = self.lib.s_csydec.idb
+        if sid == 0 or sid > idb.size - 1:
+            return
+        if stable:
+            info(
+                5, 'defining as stable particle pdgid/sid = {0}/{1}'.format(
+                    pdgid, sid))
+            idb[sid - 1] = -np.abs(idb[sid - 1])
         else:
-            self.pz = lib.s_plist.p[:npart, 2][stable]
-
-
-#=========================================================================
-# set_stable
-#=========================================================================
-def set_stable(lib, decay_mode, dbg=True):
-    idb = lib.s_csydec.idb
-
-    if dbg:
-        print("SophiaCascadeRun::set_stable(): Setting standard" +
-              " particles stable.")
-
-    #fast-mode particles
-    if decay_mode == 0:
-        stab = SibyllParticleTable()
-        for pdg_id in standard_particles:
-            print('stable,', pdg_id)
-            idb[i - 1] = -np.abs(idb[i - 1])
-        return
-
-    # keep muons pions, kaons
-    for i in range(4, 5 + 1):
-        idb[i - 1] = -np.abs(idb[i - 1])
-    for i in range(7, 18 + 1):
-        idb[i - 1] = -np.abs(idb[i - 1])
-    # K0 and K0-bar have to remain unstable to form K0S/L
-
-    if decay_mode <= 1:
-        return
-
-    # Decay mode 2 for generation of decay spectra (all conventional with
-    # lifetime >= K0S
-    if dbg:
-        print("SophiaCascadeRun::set_stable(): Setting conventional " +
-              "Sigma-, Xi0, Xi- and Lambda0 stable (decay mode).")
-    for i in range(36, 39 + 1):
-        idb[i - 1] = -np.abs(idb[i - 1])
-
-    if decay_mode <= 2:
-        return
-
-    # Conventional mesons and baryons
-    # keep eta, eta', rho's, omega, phi, K*
-    if dbg:
-        print("SophiaCascadeRun::set_stable(): Setting all " +
-              "conventional stable.")
-    # pi0
-    idb[6 - 1] = -np.abs(idb[6 - 1])
-    for i in range(23, 33 + 1):
-        idb[i - 1] = -np.abs(idb[i - 1])
-
-    # keep SIGMA, XI, LAMBDA
-    for i in range(34, 49 + 1):
-        idb[i - 1] = -np.abs(idb[i - 1])
-
-    if decay_mode <= 3:
-        return
-
-    # Charmed particles (only for version >= 2.2)
-    # keep all charmed
-    if dbg:
-        print("SophiaCascadeRun::set_stable(): Setting all " +
-              "conventional and charmed stable.")
-    for i in list(range(59, 61)) + list(range(71, 99 + 1)):
-        idb[i - 1] = -np.abs(idb[i - 1])
+            info(5, 'pdgid/sid = {0}/{1} allowed to decay'.format(pdgid, sid))
+            idb[sid - 1] = np.abs(idb[sid - 1])
+
+    def generate_event(self):
+        # Generate event (the final particles and their parameters) 
+        # by underlying Fortran library 
+        self.lib.eventgen(self.nucleon_code_number, 
+                self.energy_of_nucleon, 
+                self.energy_of_photon, 
+                self.angle_between_nucleon_and_photon, 0)
```

Comment:
The last 0 is actually a return parameter for which cross section has been selected. There should be an Cf2py intent(out) in eventget. You can pick up the return parameter as a class variable integer and make a property function that returns the string of what the interaction was, resonance, direct, multi-string etc. It certainly should not be an call argument.